site/content/blog/2019-06-05-ldap-nested-groups-modelling-and-representation-in-code.md

```diff
@@ -0,0 +1,119 @@
+---
+title: "LDAP Nested Groups: Modelling and Representation in Code"
+author: Martin Kraft
+date: 2019-06-05T12:00:00-04:00
+github: mkraft
+community: martin.kraft
+---
+
+## Defining Nested Groups
+
+The two main types of groups in LDAP are `groupOfNames` and `groupOfUniqueNames`. At minimum they have a `cn` (common name) attribute and can have membership attributes `member` or `uniqueMember`, respectively. 
+
+As an example, the below LDIF creates two groups: `developers` and `senior-developers`.
+
+```ldif
+dn: cn=developers,ou=groups,dc=www,dc=test,dc=com
+changetype: add
+objectclass: groupOfNames
+member: uid=miranda,ou=users,dc=www,dc=test,dc=com
+member: cn=senior-developers,ou=groups,dc=www,dc=test,dc=com
+
+dn: cn=senior-developers,ou=groups,dc=www,dc=test,dc=com
+changetype: add
+objectclass: groupOfNames
+member: uid=suzanne,ou=users,dc=www,dc=test,dc=com
+```
+
+The `developers` group has two members: a person `miranda` and another group `senior-developers`. The `senior-developers` group has a single member: a person `suzanne`. 
+
+When a group has another group as a member we call it a "nested group".
+
+## Objectives
+
+There are two main query operations one wants to perform when dealing with nested groups.
+
+1. List all of the members of group X.
+2. List all of the groups that person Y belongs to. 
+
+For our example, these are the correct answers we would end with:
+
+* `developers` has members `[miranda, suzanne]`
+* `senior-developers` has members `[suzanne]`
+* `suzanne` is a member of groups `[developers, senior-developers]`
+* `miranda` is a member of groups `[developers]`
+
+## Native Solution
+
+There's an "extensible match operator" called `LDAP_MATCHING_RULE_IN_CHAIN` that, if installed, traveses nested groups. It's used by adding the string `:1.2.840.113556.1.4.1941:` to your query filter.
+
+For example, this filter retrieves all of the recursive members of `developers`:
+
+```
+(&
+    (objectClass=person)
+    (memberOf:1.2.840.113556.1.4.1941:=cn=developers,ou=groups,dc=www,dc=test,dc=com)
+)
+```
+
+And this filter retrieves all of the recursive groups that `suzanne` belongs to:
+
+```
+(member:1.2.840.113556.1.4.1941:=uid=suzanne,ou=users,dc=www,dc=test,dc=com)
+```
+
+However, not all LDAP implementations natively support `LDAP_MATCHING_RULE_IN_CHAIN`, so when we set out to support LDAP nested groups in [Mattermost](https://mattermost.com/) I realized we would need to do the group traversal in our application code.
+
+## Modelling the Problem
+
+We can represent our example arrangement with a directed graph (aka digraph) as seen in *figure 1*.
+
+![nested group](/img/nested-group.png)
+*figure 1*
+
+The digraph makes the structure easy to reason about and, moreover, because cycles are possible in LDAP nested groups — group A can have group B as a member which in turn has group A as a member — the digraph is critical to avoiding needless code complexity and errors. Here's how it works...
```

Comment:
```suggestion
The digraph makes the structure easy to reason about and, moreover, because cycles are possible in LDAP nested groups — group A can have group B as a member which in turn has group A as a member — the digraph is critical to avoiding needless code complexity and errors. Here's how it works:
```